### DISCUSSION TITLE
Need Guidance: Cross-Compiling MPV on Ubuntu 24.04

### DISCUSSION THREAD

**Discussion by SiddaramMavinamar** [mpv-player/mpv#14275] [Q&A]
### Important Information

- mpv version: v0.38.0
- Platform version: Ubuntu 24.04
- Meson version: 1.3.2

### Reproduction Steps

git clone https://github.com/mpv-player/mpv.git
cd mpv
meson setup build --cross-file crossfile.txt -Dlibmpv=true
meson compile -C build

got error in compilation steps

### Expected Behavior

The build process should complete successfully with all dependencies properly resolved and compiled, resulting in a functional cross-compiled version of MPV.

### Actual Behavior

Managed to start the build but faced a new error with missing features.h. and so on

### Log File

![image](https://github.com/mpv-player/mpv/assets/142959336/6aac2223-dcf8-4bec-b574-900a12493fb0)
![image](https://github.com/mpv-player/mpv/assets/142959336/6e091776-f9c0-4992-8ddb-9b20c2597757)
![image](https://github.com/mpv-player/mpv/assets/142959336/bed96d30-442f-4854-a8e4-040f21a7195d)

I would appreciate detailed steps or documentation on how to:
Set up the cross-compilation environment for MPV on Ubuntu 24.04.
Install and configure all necessary tools and dependencies.
if any suggest 

NOTE : Successfully built MPV natively on Linux. but looking for Need to set -Dgpl=false for the Windows build.

crossfieltext.txt
c = 'x86_64-w64-mingw32-gcc'
cpp = 'x86_64-w64-mingw32-g++'
ar = 'x86_64-w64-mingw32-ar'
strip = 'x86_64-w64-mingw32-strip'
pkg-config = 'pkg-config'
exe_wrapper = 'wine64'
windres = '/usr/bin/x86_64-w64-mingw32-windres'

[host_machine]
system = 'windows'
cpu_family = 'x86_64'
cpu = 'x86_64'
endian = 'little'

### I carefully read all instruction and confirm that I did the following:

- [X] I tested with the latest mpv version to validate that the issue is not already fixed.
- [X] I provided all required information including system and mpv version.
- [X] I produced the log file with the exact same set of parameters, and conditions used in "Reproduction Steps".
- [X] I produced the log file while the behaviors described in "Actual Behavior" were actively observed.
- [X] I attached the full, untruncated log file.

**Comment by kasper93** [comment]
I've updated docs to reflect better current state of things, see [compile-windows.md](https://github.com/kasper93/mpv/blob/win_docs/DOCS/compile-windows.md).

As for your issue. Remove `pkg-config` from cross file. Or configure `pkg-config` correctly for your sysroot. By default it will detect libraries from your build machine, which is not correct for cross-compilation.

See https://mesonbuild.com/Cross-compilation.html#defining-the-environment for details, specifically `pkg_config_libdir` and `sys_root`.

**Comment by SiddaramMavinamar** [comment]
> I've updated docs to reflect better current state of things, see [compile-windows.md](https://github.com/kasper93/mpv/blob/win_docs/DOCS/compile-windows.md).
> 
> As for your issue. Remove `pkg-config` from cross file. Or configure `pkg-config` correctly for your sysroot. By default it will detect libraries from your host system, which is not correct for cross-compilation.
> 
> See https://mesonbuild.com/Cross-compilation.html#defining-the-environment for details, specifically `pkg_config_libdir` and `sys_root`.

I have changed my pkg-config path to mingw pkg but getting error like 
Not found libavcodec and remaining libraries

It seems I have to write my own pc file for mingw which is not feasible for me since am totally new to linux and same time am working on mys2 but able to generate mpv. Exe but that's not static linked 

Thanks please suggest better way

**Comment by kasper93** [comment]
>  Not found libavcodec and remaining libraries

Yes, you need to compile everything yourself when cross-compiling so the libraries are available. Meson can do it for you.

> It seems I have to write my own pc file for mingw which is not feasible for me since am totally new to linux and same time am working on mys2 but able to generate mpv. Exe but that's not static linked

You don't have to write own .pc, they would be generate when ffmpeg is installed for example to custom cross-compiling root.

> Thanks please suggest better way

Please refer to [compile-windows.md](https://github.com/kasper93/mpv/blob/win_docs/DOCS/compile-windows.md) I linked before. It should have all information you need. If something is unclear or missing, please let me know, I will update the document.

Generally it is possible to create static library in any build environment, whether it is MSYS2, cross-compilation from Linux or MSVC Clang build. There are minor differences in setup process, but this is all Meson doing heavy lifting.

**Reply by SiddaramMavinamar** [reply]
> > Not found libavcodec and remaining libraries
> 
> Yes, you need to compile everything yourself when cross-compiling so the libraries are available. Meson can do it for you.
> 
> > It seems I have to write my own pc file for mingw which is not feasible for me since am totally new to linux and same time am working on mys2 but able to generate mpv. Exe but that's not static linked
> 
> You don't have to write own .pc, they would be generate when ffmpeg is installed for example to custom cross-compiling root.
> 
> > Thanks please suggest better way
> 
> Please refer to [compile-windows.md](https://github.com/kasper93/mpv/blob/win_docs/DOCS/compile-windows.md) I linked before. It should have all information you need. If something is unclear or missing, please let me know, I will update the document.
> 
> Generally it is possible to create static library in any build environment, whether it is MSYS2, cross-compilation from Linux or MSVC Clang build. There are minor differences in setup process, but this is all Meson doing heavy lifting.

Hi @kasper93 ,

Thank you for your suggestion! I have successfully created the MPV player and managed to launch it. However, I'm facing an issue when invoking the player from a WPF application; the app crashes. Interestingly, the application does not crash when using another MPV player, which leads me to suspect either memory consumption issues or missing dependencies.

From my observation, I found that the Vulkan header might be a potential issue. Below are the details:
libplacebo| subprojects/libplacebo/src/meson.build:179: WARNING: Building without any GLSL compiler (shaderc, glslang), but with APIs required that require one (vulkan, d3d11). This build is very likely to be very limited in functionality! (for more info attached Screenshots)

d3d11                 : NO
javascript            : YES
libmpv                : YES
lua                   : NO
opengl                : YES
vulkan                : NO
wayland               : NO
x11                   : NO

Here is my cross file where I need to add the windres and dlltool configuration:
[binaries]
c = 'x86_64-w64-mingw32-gcc'
cpp = 'x86_64-w64-mingw32-g++'
![Screenshot from 2024-06-04 12-12-01](https://github.com/mpv-player/mpv/assets/142959336/f2133209-d443-4080-9beb-e32bbff16780)

ar = 'x86_64-w64-mingw32-ar'
strip = 'x86_64-w64-mingw32-strip'
windres = 'x86_64-w64-mingw32-windres'
dlltool = 'x86_64-w64-mingw32-dlltool'
exe_wrapper = '/usr/lib/wine/wine64'

[properties]
needs_exe_wrapper = true

[host_machine]
system = 'windows'
cpu_family = 'x86_64'
cpu = 'x86_64'
endian = 'little'

Could you please advise if I missed any dependencies or configurations that might be causing this issue?

Thanks in advance for your help!

![Screenshot from 2024-06-04 12-12-19](https://github.com/mpv-player/mpv/assets/142959336/47d0f34c-c496-4012-98c7-e9abeaf5d7fa)

**Reply by kasper93** [reply]
> the app crashes

Is not very specific, you have to debug your application and fix the issues.

> WARNING: Building without any GLSL compiler (shaderc, glslang), but with APIs required that require one (vulkan, d3d11). This build is very likely to be very limited in functionality!

Well, yes. You have to compile and link shaderc and spriv-cross for d3d11 and Vulkan support.

**Reply by SiddaramMavinamar** [reply]
> > the app crashes
> 
> Is not very specific, you have to debug your application and fix the issues.

**Regarding my previous question, to clarify, I am able to launch MPV player from this link: https://sourceforge.net/projects/mpv-player-windows/files/. However, I am unable to achieve the same result when building from source, leading me to wonder if the numerous warnings during the build process might be the cause.**
> 
> > WARNING: Building without any GLSL compiler (shaderc, glslang), but with APIs required that require one (vulkan, d3d11). This build is very likely to be very limited in functionality!
> 
> Well, yes. You have to compile and link shaderc and spriv-cross for d3d11 and Vulkan support.

As you suggested, I need to compile and link shaderc and spriv-cross for d3d11 and Vulkan support. I have done the following steps and compiled, though I got these errors and warnings:

i have followed this but not bale to compile 

meson wrap install vulkan-headers 
meson wrap install lua
meson wrap install luajit 

WARNING: Building without any GLSL compiler (shaderc, glslang), but with APIs required that require one (vulkan, d3d11). This build is very likely to be very limited in functionality!

harfbuzz| WARNING: Project specifies a minimum meson_version '>= 0.55.0' but uses features which were added in newer versions:
harfbuzz| * 1.3.0: {'fs.relative_to'}
harfbuzz| Subproject harfbuzz finished.
libass| subprojects/libass/meson.build:251: WARNING: nasm was not found; ASM functions are disabled. Install nasm >= 2.10 for a significantly faster libass build.
libplacebo| WARNING: Library target 'placebo' has 'name_prefix' set. Compilers may not find it from its '-lplacebo' linker flag in the 'libplacebo.pc' pkg-config file.
libplacebo| WARNING: Library target 'placebo' has 'name_prefix' set. Compilers may not find it from its '-lplacebo-uninstalled.pc' pkg-config file.

Unicode callbacks (you want at least one)
    Builtin                   : YES
    Glib                      : NO
    ICU                       : YES

  Font callbacks (the more the merrier)
    Builtin                   : YES
    FreeType                  : NO

  Dependencies used for command-line utilities
    Cairo                     : NO
    Chafa                     : NO

  Additional shapers
    Graphite2                 : NO
    WebAssembly (experimental): NO

  Platform shapers (not normally needed)
    CoreText                  : NO
    DirectWrite (experimental): NO
    GDI/Uniscribe             : NO

  Other features
    Documentation             : NO
    GObject bindings          : NO
    Cairo integration         : NO
    Introspection             : NO
    Experimental APIs         : NO

  Testing
    Tests                     : NO
    Benchmark                 : NO

libass 0.17.2

    Font providers   : DirectWrite and GDI (Win32)
    ASM optimizations: NO

libplacebo 7.349.0

  Optional features
    d3d11       : NO
    dovi        : YES
    gl-proc-addr: YES
    glslang     : NO
    lcms        : YES
    libdovi     : NO
    opengl      : YES
    shaderc     : NO
    vk-proc-addr: NO
    vulkan      : YES
    xxhash      : NO

mpv 0.38.0-UNKNOWN

    d3d11                 : NO
    javascript            : YES
    libmpv                : YES
    lua                   : NO
    opengl                : YES
    vulkan                : NO
    wayland               : NO
    x11                   : NO
    
     
  
I would like to express my gratitude for your assistance so far. sorry for asking more question and thanks for your help"
    Is there any better way to resolve these warning or build dependence please assist me like below you have donehere https://github.com/kasper93/mpv/blob/win_docs/DOCS/compile-windows.md


![Screenshot from 2024-06-04 14-58-52](https://github.com/mpv-player/mpv/assets/142959336/b652d867-fc87-41b2-af0d-f860e4620042)

**Comment by SiddaramMavinamar** [comment]
Hi @kasper93 ,
Thanks to your help, I've managed to integrate the mpv media player into my WPF app seamlessly, without requiring any fixes or alterations. The instructions provided here to be sufficient to build MPV
https://github.com/kasper93/mpv/blob/win_docs/DOCS/compile-windows.md

However, I encountered an issue while trying to pass certain arguments. Specifically, I received an error stating: "**Error parsing option no-osc (option not found)**"
To address this, I need to either remove the no-osc option from the arguments or find an alternative solution to achieve the desired functionality ( **I mean while building MPV if any please suggest     Lua dependency installed but not found while building**).

 string mpvArgs = "\"" + rel_video_path + "\"" +
 " --force-window=yes" +
 " --fullscreen" +
 " **--no-osc** --no-osd-bar" +
 " --hwdec=auto-safe" +
 " --input-media-keys=no " +
 "  --no-input-terminal " +
 " --no-taskbar-progress " +
 " --no-input-cursor " +
 "--video-unscaled=no" +
 " --loop=inf --keep-open";

**Comment by rboxeur** [comment]
I worked on a project like this one a couple years ago. Outdated documentation https://github.com/rboxeur/mpv-i686-cross-compiling-MinGW32-Doc/blob/master/index.rst
BUT here is the guid I followed
https://github.com/qyot27/mpv/blob/extra-new/DOCS/crosscompile-mingw-tedious.txt

**Reply by SiddaramMavinamar** [reply]
> I worked on a project like this one a couple years ago. Outdated documentation https://github.com/rboxeur/mpv-i686-cross-compiling-MinGW32-Doc/blob/master/index.rst BUT here is the guid I followed https://github.com/qyot27/mpv/blob/extra-new/DOCS/crosscompile-mingw-tedious.txt


Hi, as I mentioned earlier, I was able to build the mpv player with @kasper93 suggestions and help. He shared the link 

https://github.com/kasper93/mpv/blob/win_docs/DOCS/compile-windows.md

which I feel is better than all other solutions. However, the issue now is that I am unable to add Lua libraries via subproject or other means. I am looking for a solution to add Vulkan Lua Luajit libray

**Reply by kasper93** [reply]
for luajit, you can from root mpv directory do `meson wrap install luajit` and it will be able to use it. For Vulkan, you have to compile shaderc externally. There are way to do it with meson, but I wouldn't recommend it, but if you look at `ci/build-win32.ps1` there is working example of that.

**Reply by SiddaramMavinamar** [reply]
> for luajit, you can from root mpv directory do `meson wrap install luajit` and it will be able to use it. For Vulkan, you have to compile shaderc externally. There are way to do it with meson, but I wouldn't recommend it, but if you look at `ci/build-win32.ps1` there is working example of that.


For LuaJIT and even Lua, I used meson wrap install, but it's failing to install dependencies. As I found, the wrap files are in the subprojects folder but weren't extracted and installed. I got the following error:
meson.build:732:5: ERROR: Problem encountered: Lua enabled but no suitable Lua version could be found!
I did the following:
meson wrap install vulkan-headers
meson wrap install lua
meson wrap install luajit

These above files are already installed in root /usr/bin but it's not working for cross compile

Any suggestions / idea you have used, if you know how to add like ffmpeg or libplacebo kind

**Reply by kasper93** [reply]
`meson wrap install luajit` and `meson configure -Dlua=luajit` should work. Generally it is like puzzles, you need to figure out how to build everything, there is not much we can do to help, you have all information needed. Also you could try https://github.com/shinchiro/mpv-winbuild-cmake which is all-in-one solution. Beware they are building GPL builds by default.

**Reply by SiddaramMavinamar** [reply]
> `meson wrap install luajit` and `meson configure -Dlua=luajit` should work. Generally it is like puzzles, you need to figure out how to build everything, there is not much we can do to help, you have all information needed. Also you could try https://github.com/shinchiro/mpv-winbuild-cmake which is all-in-one solution. Beware they are building GPL builds by default.

Okay, thank you for your kind help; your solution is the best. I also tried this all-in-one solution: https://github.com/shinchiro/mpv-winbuild-cmake. However, I feel it doesn't provide much information.

Anyhow, I am close to meeting my requirements. The reason I'm focusing on Lua is that I am not able to pass the **--no-osc** argument. Enabling Lua with -**Dlua=enabled** results in the error mentioned above.

**Reply by kasper93** [reply]
Try `-Dlua=luajit`, to force specific version.

**Reply by SiddaramMavinamar** [reply]
Thank you 
Let me try and update you the same

**Comment by Andarwinux** [comment]
Just use [mpv-winbuild](https://github.com/zhongfly/mpv-winbuild) to build mpv via GHA. you can find prebuilt LGPL builds.

**Reply by SiddaramMavinamar** [reply]
> Just use [mpv-winbuild](https://github.com/zhongfly/mpv-winbuild) to build mpv via GHA. you can find prebuilt LGPL builds.

Tried this no proper document with given link when trying to build it has error 
./build.sh

**Reply by zhongfly** [reply]
If you want a lgpl(gpl=false) libmpv, go to release and download the files with 'lgpl'
![Screenshot](https://github.com/mpv-player/mpv/assets/11155705/a62c7902-9989-45ec-a3bf-503d4d568b54)

**Reply by zhongfly** [reply]
If you need a mpv.exe under lgpl license , that doesn't make any sense. From what I know of the gpl license, a non-GPL application can call a GPL executable at the command line

**Reply by SiddaramMavinamar** [reply]
Thank you for your help. I need to build mpv.exe with **-Dgpl=false**. I've tried every possible way, and while the method in this link 
 - https://github.com/mpv-player/mpv/blob/master/DOCS/compile-windows.md suggested was helpful, I encountered an issue with Lua not compiling, resulting in the error: "Error parsing option no-osc (option not found)".
 
 The build from the link -https://sourceforge.net/projects/mpv-player-windows/files/
 you provided works fine, except that I need to compile with -Dgpl=false.
 
 any input please

**Reply by zhongfly** [reply]
I think you didn't know what's meaning of lgpl or -Dgpl=false, and https://sourceforge.net/projects/mpv-player-windows/files/ is not my build.

**Reply by SiddaramMavinamar** [reply]
General Public License, a common free software license. Setting -Dgpl=false can be used to signify that GPL-licensed components should not be used or that certain features contingent on GPL code should be disabled.

could you please explain how to build full pledged MVP with  -Dgpl=false
